### ISSUE TITLE
Exception on startup

### ISSUE THREAD

**Issue by RoiEXLab** [triplea-game/triplea#2584]
I'm not sure what exactly is causing this, but I think this regression has been introduced in #2573 
I can no longer reproduce this since I explicitly loaded a map, but could reproduce it consistenly before.
```
triplea.engine.version.bin:1.9
java.lang.IllegalStateException
	at com.google.common.base.Preconditions.checkState(Preconditions.java:485)
	at games.strategy.engine.framework.ui.background.BackgroundTaskRunner.runInBackgroundAndReturn(BackgroundTaskRunner.java:103)
	at games.strategy.engine.framework.ui.background.BackgroundTaskRunner.runInBackgroundAndReturn(BackgroundTaskRunner.java:71)
	at games.strategy.engine.framework.ui.GameChooserModel.newInstance(GameChooserModel.java:66)
	at games.strategy.engine.framework.startup.mc.GameSelectorModel.selectByName(GameSelectorModel.java:298)
	at games.strategy.engine.framework.startup.mc.GameSelectorModel.loadDefaultGame(GameSelectorModel.java:282)
	at games.strategy.engine.framework.GameRunner.lambda$7(GameRunner.java:278)
	at games.strategy.engine.framework.ui.background.BackgroundTaskRunner.lambda$0(BackgroundTaskRunner.java:43)
	at games.strategy.engine.framework.ui.background.BackgroundTaskRunner.lambda$1(BackgroundTaskRunner.java:71)
	at games.strategy.engine.framework.ui.background.BackgroundTaskRunner$1.doInBackground(BackgroundTaskRunner.java:114)
	at javax.swing.SwingWorker$1.call(SwingWorker.java:295)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at javax.swing.SwingWorker.run(SwingWorker.java:334)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
java.lang.IllegalStateException
	at com.google.common.base.Preconditions.checkState(Preconditions.java:485)
	at games.strategy.engine.framework.ui.background.BackgroundTaskRunner.runInBackgroundAndReturn(BackgroundTaskRunner.java:103)
	at games.strategy.engine.framework.ui.background.BackgroundTaskRunner.runInBackgroundAndReturn(BackgroundTaskRunner.java:71)
	at games.strategy.engine.framework.ui.GameChooserModel.newInstance(GameChooserModel.java:66)
	at games.strategy.engine.framework.startup.mc.GameSelectorModel.selectByName(GameSelectorModel.java:298)
	at games.strategy.engine.framework.startup.mc.GameSelectorModel.loadDefaultGame(GameSelectorModel.java:282)
	at games.strategy.engine.framework.GameRunner.lambda$7(GameRunner.java:278)
	at games.strategy.engine.framework.ui.background.BackgroundTaskRunner.lambda$0(BackgroundTaskRunner.java:43)
	at games.strategy.engine.framework.ui.background.BackgroundTaskRunner.lambda$1(BackgroundTaskRunner.java:71)
	at games.strategy.engine.framework.ui.background.BackgroundTaskRunner$1.doInBackground(BackgroundTaskRunner.java:114)
	at javax.swing.SwingWorker$1.call(SwingWorker.java:295)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at javax.swing.SwingWorker.run(SwingWorker.java:334)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by ssoloff** [comment]
Weird, I just saw it, too, while testing some changes to #2581.  I did not observe this error yesterday, and I haven't rebased since then (that branch is currently based off of 4899e8325 (#2572)).

**Comment by ssoloff** [comment]
But it appears to be due to what I mentioned earlier.  Namely, that there is a recursive call to `BackgroundTaskRunner` while loading the default game.  A quick, but smelly, fix seems to be the following:

```diff
diff --git a/src/main/java/games/strategy/engine/framework/ui/GameChooserModel.java b/src/main/java/games/strategy/engine/framework/ui/GameChooserModel.java
index cca8964ef..48e98455c 100644
--- a/src/main/java/games/strategy/engine/framework/ui/GameChooserModel.java
+++ b/src/main/java/games/strategy/engine/framework/ui/GameChooserModel.java
@@ -23,6 +23,7 @@ import java.util.zip.ZipFile;
 
 import javax.swing.DefaultListModel;
 import javax.swing.JOptionPane;
+import javax.swing.SwingUtilities;
 
 import org.xml.sax.SAXException;
 import org.xml.sax.SAXParseException;
@@ -63,9 +64,12 @@ public class GameChooserModel extends DefaultListModel<GameChooserEntry> {
    *         Thread than the Swing Event Dispatch Thread.
    */
   public static GameChooserModel newInstance() throws InterruptedException {
-    return new GameChooserModel(BackgroundTaskRunner.runInBackgroundAndReturn(
-        "Loading all available games...",
-        GameChooserModel::parseMapFiles));
+    final Set<GameChooserEntry> gameChooserEntries = SwingUtilities.isEventDispatchThread()
+        ? BackgroundTaskRunner.runInBackgroundAndReturn(
+            "Loading all available games...",
+            GameChooserModel::parseMapFiles)
+        : parseMapFiles();
+    return new GameChooserModel(gameChooserEntries);
   }
 
   @Override
```

**Comment by ssoloff** [comment]
You can trigger the code path that raises the exception by deleting the `games/strategy/triplea/settings/DEFAULT_GAME_URI_PREF` user preference.

**Comment by ssoloff** [comment]
Here's another possible solution that puts the responsibility of calling `GameChooserModel#parseMapFiles()` on the caller so they can decide if it needs to be done on the current thread or a background thread (this is a diff with whitespace changes ignored):

```diff
diff --git a/src/main/java/games/strategy/engine/framework/startup/mc/GameSelectorModel.java b/src/main/java/games/strategy/engine/framework/startup/mc/GameSelectorModel.java
index d3315e1c3..3c7962275 100644
--- a/src/main/java/games/strategy/engine/framework/startup/mc/GameSelectorModel.java
+++ b/src/main/java/games/strategy/engine/framework/startup/mc/GameSelectorModel.java
@@ -294,8 +294,7 @@ public class GameSelectorModel extends Observable {
     }
     final String userPreferredDefaultGameName = ClientSetting.DEFAULT_GAME_NAME_PREF.value();
 
-    try {
-      final GameChooserModel model = GameChooserModel.newInstance();
+    final GameChooserModel model = new GameChooserModel(GameChooserModel.parseMapFiles());
     GameChooserEntry selectedGame = model.findByName(userPreferredDefaultGameName);
     if (selectedGame == null) {
       selectedGame = model.findByName(userPreferredDefaultGameName);
@@ -318,9 +317,5 @@ public class GameSelectorModel extends Observable {
       }
     }
     return selectedGame;
-    } catch (final InterruptedException e) {
-      Thread.currentThread().interrupt();
-      return null;
-    }
   }
 }
diff --git a/src/main/java/games/strategy/engine/framework/ui/GameChooser.java b/src/main/java/games/strategy/engine/framework/ui/GameChooser.java
index 733cc2fcc..ea63cefe8 100644
--- a/src/main/java/games/strategy/engine/framework/ui/GameChooser.java
+++ b/src/main/java/games/strategy/engine/framework/ui/GameChooser.java
@@ -24,6 +24,7 @@ import javax.swing.JSplitPane;
 import javax.swing.SwingUtilities;
 
 import games.strategy.engine.data.GameData;
+import games.strategy.engine.framework.ui.background.BackgroundTaskRunner;
 import games.strategy.util.LocalizeHtml;
 
 public class GameChooser extends JDialog {
@@ -99,7 +100,10 @@ public class GameChooser extends JDialog {
 
   public static GameChooserEntry chooseGame(final Frame parent, final String defaultGameName)
       throws InterruptedException {
-    final GameChooser chooser = new GameChooser(parent, GameChooserModel.newInstance());
+    final GameChooserModel gameChooserModel = new GameChooserModel(BackgroundTaskRunner.runInBackgroundAndReturn(
+        "Loading all available games...",
+        GameChooserModel::parseMapFiles));
+    final GameChooser chooser = new GameChooser(parent, gameChooserModel);
     chooser.setSize(800, 600);
     chooser.setLocationRelativeTo(parent);
     chooser.selectGame(defaultGameName);
diff --git a/src/main/java/games/strategy/engine/framework/ui/GameChooserModel.java b/src/main/java/games/strategy/engine/framework/ui/GameChooserModel.java
index f2d4595f2..90bbacba3 100644
--- a/src/main/java/games/strategy/engine/framework/ui/GameChooserModel.java
+++ b/src/main/java/games/strategy/engine/framework/ui/GameChooserModel.java
@@ -44,30 +44,10 @@ public class GameChooserModel extends DefaultListModel<GameChooserEntry> {
     SUCCESS, ERROR
   }
 
-  @VisibleForTesting
-  GameChooserModel(final Set<GameChooserEntry> gameChooserEntries) {
+  public GameChooserModel(final Set<GameChooserEntry> gameChooserEntries) {
     gameChooserEntries.stream().sorted().forEach(this::addElement);
   }
 
-  /**
-   * Creates a new Instance of a GameChooserModel.
-   * Must be called on the EDT, will not return until all
-   * map files have been scanned, but is not going to
-   * effectively block the EDT.
-   *
-   * @return The newly created instance.
-   *
-   * @throws InterruptedException If this Thread is being interrupted
-   *         while waiting for the SwingWorker to complete.
-   * @throws IllegalStateException If this method is called on any other
-   *         Thread than the Swing Event Dispatch Thread.
-   */
-  public static GameChooserModel newInstance() throws InterruptedException {
-    return new GameChooserModel(BackgroundTaskRunner.runInBackgroundAndReturn(
-        "Loading all available games...",
-        GameChooserModel::parseMapFiles));
-  }
-
   @Override
   public GameChooserEntry get(final int i) {
     return super.get(i);
@@ -85,7 +65,7 @@ public class GameChooserModel extends DefaultListModel<GameChooserEntry> {
     return Arrays.asList(files);
   }
 
-  private static Set<GameChooserEntry> parseMapFiles() {
+  public static Set<GameChooserEntry> parseMapFiles() {
     final List<File> files = allMapFiles();
     final Set<GameChooserEntry> parsedMapSet = new HashSet<>(files.size());
     final ExecutorService service = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors() + 2);
```

**Comment by panther2** [comment]
FWIW I can constantly reproduce this error with 7469.
It occurs not only when starting TripleA but also when leaving a game with "File - Leave Game".

**Comment by ssoloff** [comment]
Running `GameSelectorModelTest`, as described in #2588, is another way to trigger this error.

**Comment by ssoloff** [comment]
A temporary fix is now available in 1.9.0.0.7475.

We probably still need to discuss this a bit more to determine if there's a better solution before closing this issue.  I don't like the fact that `newInstance()` now may throw `InterruptedException` depending on which thread it's invoked upon.  That forces callers who are calling it off of the EDT to handle `InterruptedException` even though it will never be thrown.

**Comment by ssoloff** [comment]
@RoiEXLab Do you have any suggestions for improving the **second** solution proposed [above](https://github.com/triplea-game/triplea/issues/2584#issuecomment-343023829)?  As I mentioned in #2587, I don't like the current fix because it forces callers to handle `InterruptedException` even when it can't possibly be thrown.  Without any better ideas, I'll probably submit a PR fairly close to the second proposal.

**Comment by RoiEXLab** [comment]
@ssoloff I'm fine with your second proposal, perhaps you should create an overloaded constructor for gamechoosermodel which is required to be called on the EDT and parses the map files without having to explicitly pass them.
But I'd be fine either way.

**Comment by ssoloff** [comment]
Fixed in 1.9.0.0.7501 via #2593.
